packages/mds-ingest-service/repository/index.ts

```
@@ -14,14 +14,222 @@
+          qb => {
+            return qb
+              .select(
+                'device_id, id as event_id, RANK() OVER (PARTITION BY device_id ORDER BY timestamp DESC) AS rownum'
+              )
+              .from(EventEntity, 'e')
+              .where('timestamp >= :start AND timestamp <= :end', { start, end })
+          },
+          'last_device_event',
+          'last_device_event.event_id = events.id AND last_device_event.rownum = 1'
+        )
+      }
+
+      if (grouping_type === 'latest_per_trip') {
+        query.innerJoin(
+          qb => {
+            return qb
+              .select('trip_id, id as event_id, RANK() OVER (PARTITION BY trip_id ORDER BY timestamp DESC) AS rownum')
+              .from(EventEntity, 'e')
+              .where('timestamp >= :start AND timestamp <= :end', { start, end })
+          },
+          'last_trip_event',
+          'last_trip_event.event_id = events.id AND last_trip_event.rownum = 1'
+        )
+      }
+
+      if (event_types) {
+        query.andWhere('events.event_types && :event_types', { event_types })
+      }
+
+      if (propulsion_types) {
+        query.andWhere('devices.propulsion_types && :propulsion_types', { propulsion_types })
+      }
+
+      if (device_ids) {
+        query.andWhere('events.device_id = ANY(:device_ids)', { device_ids })
+      }
+
+      if (vehicle_types) {
+        query.andWhere('devices.vehicle_type = ANY(:vehicle_types)', { vehicle_types })
+      }
+
+      if (vehicle_states) {
+        query.andWhere('events.vehicle_state = ANY(:vehicle_states)', { vehicle_states })
+      }
+
+      if (device_or_vehicle_id) {
+        if (isUUID(device_or_vehicle_id)) {
+          query.andWhere('devices.device_id = :device_or_vehicle_id', { device_or_vehicle_id })
+        } else {
+          query.andWhere('devices.vehicle_id = :device_or_vehicle_id', { device_or_vehicle_id })
+        }
+      }
+
+      if (provider_ids && provider_ids.every(isUUID)) {
```

It seems like the `isUUID` check should have been verified in a validator long before we get here.

---

It's an important detail to validate here, b/c Postgres will _not_ compare any UUID fields to values that are not also UUIDs, it raises a SQL exception if you happen to feed it something like `XXYY123`.

---

Fair, but it would probably be better to throw an error or filter out the invalid ones rather than ignore everything in the presence of one non-uuid.

---

added a SchemaValidator!